src/platforms/linux/xdgportal.cpp

```diff
@@ -108,3 +112,131 @@ QString XdgPortal::parentWindow() {
+                           "/org/freedesktop/systemd1",
+                           "org.freedesktop.systemd1.Manager");
+
+  QDBusMessage getunit = sdManager.call("GetUnitByPID", ownPid);
+  if (getunit.type() == QDBusMessage::ErrorMessage) {
+    logger.warning() << "Failed to get scope:" << getunit.errorMessage();
+    return;
+  }
+  QList<QVariant> result = getunit.arguments();
+  if ((getunit.type() != QDBusMessage::ReplyMessage) || result.isEmpty()) {
+    logger.warning() << "Bad reply for current scope:";
+    return;
+  }
+
+  // Fetch the names of the unit to figure out the appid.
+  QDBusObjectPath unitPath = result.first().value<QDBusObjectPath>();
+  QDBusInterface ownUnit("org.freedesktop.systemd1", unitPath.path(),
+                         "org.freedesktop.systemd1.Unit");
+  QStringList unitNames = ownUnit.property("Names").toStringList();
+
+  // Use the D-Bus object path as a fallback if there are no names.
+  unitNames.append(decodeSystemdEscape(unitPath.path().split('/').last()));
+
+  // Check to see if we have a valid application scope.
+  QString unitAppId;
+  for (const QString& name : unitNames) {
+    if (!name.endsWith(".scope")) {
+      continue;
+    }
+    QStringList scopeSplit = name.first(name.size() - 6).split('-');
+    if (scopeSplit[0] != "app") {
+      continue;
+    }
+
+    // Remove the last element of the scope if it's a number.
+    bool isDigit = false;
+    scopeSplit.last().toULong(&isDigit);
+    if (isDigit) {
+      scopeSplit.removeLast();
+    }
+
+    // The appId should be the last remaining element of the scope.
+    // If we found an appId then we have no work to do.
+    logger.info() << "Launched as app" << scopeSplit.last();
+    return;
+  }
+
+  // Request a new scope from systemd via D-Bus
+  QString newScopeName =
+      QString("app-%1-%2.scope").arg(appId, QString::number(getpid()));
+  SystemdUnitPropList properties;
+  SystemdUnitAuxList aux;
+  QList<uint> pidlist({ownPid});
+  properties.append(SystemdUnitProp("PIDs", QVariant::fromValue(pidlist)));
+
+  logger.debug() << "Creating scope:" << newScopeName;
+  QDBusMessage msg =
+      sdManager.call("StartTransientUnit", newScopeName, "fail",
+                     QVariant::fromValue(properties), QVariant::fromValue(aux));
+  if (msg.type() == QDBusMessage::ErrorMessage) {
+    logger.warning() << "Failed to create scope:" << msg.errorMessage();
+    return;
+  }
+  QDBusObjectPath jobPath = result.first().value<QDBusObjectPath>();
+  if ((msg.type() != QDBusMessage::ReplyMessage) || jobPath.path().isEmpty()) {
+    logger.warning() << "Bad reply for create scope";
+    return;
+  }
+
+  // And now for the gross part. StartTransientUnit() returns a systemd job that
+  // will create our scope for us, and move us into it. But we have to wait for
+  // that job to finish before we can proceed. However, we can't do this async
+  // because we don't have a QCoreApplication setup yet. The only way I can
```

Comment:
I think this is is fine, we propbably could create the qApp as first thing but then again if we have this async task to move the scope.  Settingsholder would need to wait for that ... we would need to touch a lot of places to work with a `Future<Settingholder*>` so i am fine with polling, unless you think that is too long :)

Reply:
I don't think it's worth the effort. The job seems to finish in tens of milliseconds and it's almost always done in the first iteration of the loop. We mostly just need it to ensure we're not in a some kind of race condition when `SettingsHolder` starts up.